src/vdaf/mastic.rs

```diff
@@ -379,32 +382,418 @@ where
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match prep_state.szk_query_state {
+            Some(_) => SzkJointShare::<SEED_SIZE>::decode_with_param(&true, bytes),
+            None => SzkJointShare::<SEED_SIZE>::decode_with_param(&false, bytes),
+        }
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Aggregator<SEED_SIZE, NONCE_SIZE> for Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    type PrepareState = MasticPrepareState<T::Field, SEED_SIZE>;
+    type PrepareShare = MasticPrepareShare<T::Field, SEED_SIZE>;
+    type PrepareMessage = MasticPrepareMessage<SEED_SIZE>;
+
+    fn prepare_init(
+        &self,
+        verify_key: &[u8; SEED_SIZE],
+        agg_id: usize,
+        agg_param: &MasticAggregationParam,
+        nonce: &[u8; NONCE_SIZE],
+        public_share: &MasticPublicShare<VidpfWeight<T::Field>>,
+        input_share: &MasticInputShare<T::Field, SEED_SIZE>,
+    ) -> Result<
+        (
+            MasticPrepareState<T::Field, SEED_SIZE>,
+            MasticPrepareShare<T::Field, SEED_SIZE>,
+        ),
+        VdafError,
+    > {
+        let id = match agg_id {
+            0 => Ok(VidpfServerId::S0),
+            1 => Ok(VidpfServerId::S1),
+            _ => Err(VdafError::Uncategorized(
+                "Invalid aggregator ID".to_string(),
+            )),
+        }?;
+        let mut eval_proof = P::init(
+            verify_key,
+            &self.domain_separation_tag(DST_PATH_CHECK_BATCH),
+        );
+        let mut output_shares = Vec::<T::Field>::with_capacity(
+            self.vidpf.weight_parameter * agg_param.level_and_prefixes.prefixes().len(),
+        );
+        let mut cache_tree = BinaryTree::<VidpfEvalCache<VidpfWeight<T::Field>>>::default();
+        let cache = VidpfEvalCache::<VidpfWeight<T::Field>>::init_from_key(
+            id,
+            &input_share.vidpf_key,
+            &self.vidpf.weight_parameter,
+        );
+        cache_tree
+            .insert(Path::empty(), cache)
+            .expect("Should alwys be able to insert into empty tree at root");
+        for prefix in agg_param.level_and_prefixes.prefixes() {
+            let mut value_share = self.vidpf.eval_with_cache(
+                id,
+                &input_share.vidpf_key,
+                public_share,
+                prefix,
+                &mut cache_tree,
+                nonce,
+            )?;
+            eval_proof.update(&value_share.proof);
+            output_shares.append(&mut value_share.share.0);
+        }
+
+        let szk_verify_opt = if agg_param.require_weight_check {
+            let root_share = self.vidpf.eval_root_with_cache(
+                id,
+                &input_share.vidpf_key,
+                public_share,
+                &mut cache_tree,
+                nonce,
+            )?;
+            Some(self.szk.query(
+                root_share.as_ref(),
+                input_share.proof_share.clone(),
```

Comment:
We should be able to avoid this clone by passing a reference to `query()` rather than the proof share itself. This may be a lot of savings, as the proof share may be 100s of bytes.